### ISSUE TITLE
Soft-token MFA causes indefinite hang (and locks me out of my account)

### ISSUE THREAD

**Issue by ScottG489** [mrooney/mintapi#229]
I'm having some trouble getting the `soft-token` MFA method to work.

First, thanks a lot to @chino for adding this support.

## What worked
Next, for some background. Previous to setting up the **authenticator app MFA** method (as described below), I also had **SMS MFA** set up. I would use this command:

`mintapi --headless --credit-score '<username>' '<password>'`

After running this I would be prompted to enter the 6-digit code which was texted to me. This worked fine and I got a successful response with my credit score. Note that I'm running this in a `docker` container and I've included information about that below.

## Setting up MFA for an authenticator app
Here are the specific steps I took to set up the **authenticator app MFA** method:

- I went to https://accounts.intuit.com/app/account-manager/security
- I clicked on _"Two-step verification"_
- I clicked on _"Turn on Authenticator app"_
- I selected _"Android"_ then clicked _"Continue"_
- I took the code under _"Manual entry"_ and plugged it into the following command: `oathtool -b --totp '<manual entry code>'`
- I entered the code in the text box

After this it takes me back to the _"Two-step verification"_ page and says **"Authenticator app is ON"**.

I'm also able to go to the site in my browser and log in by using my credentials and then the 6-digit code returned to me by using the `oathtool` mentioned above. So this indicates my process for getting the token is correct. Sorry if this was a bit verbose but I wanted to be explicit. Now onto using it with the `mintapi` cli tool.

## Docker setup
I'm running this in `docker`. Here is the `Dockerfile`:
```
FROM ubuntu:20.04

RUN apt-get update && DEBIAN_FRONTEND="noninteractive" TZ="America/New_York" apt install -y git wget jq python3 python3-pip

RUN wget https://dl.google.com/linux/direct/google-chrome-stable_current_amd64.deb
RUN dpkg --install google-chrome-stable_current_amd64.deb || DEBIAN_FRONTEND="noninteractive" TZ="America/New_York" apt-get -fy install
RUN pip3 install mintapi

ENTRYPOINT ["mintapi"]
```

I then build and run this image:
```
docker build -t local/mintapi .
docker run -it --entrypoint bash local/mintapi
```
As mentioned before running `mintapi --headless --credit-score '<username>' '<password>'` in this environment works great and did so when I had only **SMS MFA** set up.

## Attempting to use the soft-token
In order to use the soft-token I run the following command (from within the container):
```
CODE=$(oathtool -b --totp '<manual entry code>' | base32)
mintapi --headless --credit-score --mfa-method soft-token --mfa-token "$CODE" '<username>' '<password>'
```
With the **menual entry code** being the one given to me as explained in the **Setting up MFA for an authenticator app** section above.

When I run this two things happen:
1. The application seems to hang indefinitely
2. I'm locked out from using **authenticator app MFA** to access my account in the browser (I still have **SMS MFA** set up and can still log in with that as a backup)

I'm guessing the latter is due to the application attempting to enter the code and failing over and over. This also seems to be somewhat confirmed from https://github.com/mrooney/mintapi/issues/137#issuecomment-683480698

Does anyone know what may be going wrong here?

**Comment by ScottG489** [comment]
I'd like to make a correction. I've temporarily disabled **authenticator app MFA** and now I'm having issues with even the **SMS MFA** even though it was working before. The command I'm running is 

`mintapi --headless --credit-score '<username>' '<password>'`

Now instead of prompting me to enter my MFA token that was sent to my SMS number it instead gives me this message:

`Giving up on handling MFA. Please complete the MFA process manually in the browser.`

One particular thing I noticed was that I start to get the above message around the time the text message arrives on my phone.

Unfortunately I'm having trouble troubleshooting this since I am not able to run without --headless on my local machine. I have the correct version of `chromedriver` to match up with my chromium install. I suspect it's possible it has something to do with the fact that it's an ubuntu snap package.

**Comment by chino** [comment]
Is this a typo or copy/paste issue here?   The quoting looks wrong and you're also not using a sub shell to capture the output correctly?   Can you print the output of `$CODE` to verify it's what you expect?

```
CODE=oathtool -b --totp '<manual entry code> | base32'
```

You really need some kind of debugging enabled.   You could try to take a screenshot of the failure and/or get non-headless version working so you can watch what's going on.  There is various ways to do it like running something like selenoid which can give you vnc and screen recordings, or a quick image with a lightweight window manager (ex: fluxbox) and x11vnc for instance, also I've seen things like running x11 apps in container against your own desktop X server.  Even if you can simply enable debug logs to see what's goin going on.

I imagine it's stuck retrying at the login screen and hence somehow locks you out. If you pass the wrong credentials I believe it does end up doing that.

I do have a little patch locally for a simple back-off which has worked just fine for me so far but I haven't pushed it anywhere yet.  I would try to get some debugging working first to see what your issue is first anyway.

**Comment by ScottG489** [comment]
Thanks for the catch, that was a typo. I corrected it in the original comment. However, I have actually printed out the value and manually passed it to `mintapi` to verify it's correct as you mentioned.

Thanks for the ideas on testing it locally and I think you're probably right that it's stuck at the login screen. I haven't ever done any X11 forwarding to a container before and was hoping I wouldn't have to get my hands dirty doing that just yet.

Also on the topic of debugging, is there a way to enable debugging log output for `mintapi`? It might not be as helpful as seeing what's happening in the browser but it would hopefully give a little insight into what's going on. If there was some kind of looping happening that might be indicated in the logs.

Also if it wouldn't be too much trouble, would you mind pushing your back-off code to a branch? I see that you have a fork and I'd be able to get it working from a branch from there. I understand it might not be solid enough to be a PR but I'd like to try it out as is if it works :)

**Comment by chino** [comment]
Sure I'll get to get to it tonight. It's just a very simple snippet.

You can run mintapi out of a clone to tweak the logging. I'm not exactly sure off hand right now as I'm not at a system.

Check out selenoid it makes it easy to record I believe. I haven't tried it my self but the free version might make it easy.

**Comment by ScottG489** [comment]
Hey @chino! Just thought I'd bump this for you in case it fell off your radar :)

**Comment by ScottG489** [comment]
Sorry to bug you again @chino, but would it be possible to push whatever you've got to a branch? Thanks :)

**Comment by chino** [comment]
It's really nothing special and I'd rather some review/thought before attempting to merge this:
https://github.com/chino/mintapi/commit/27e469c8ce849b42aecc230fe5489999bd4bb551

The real thing you should focus on is actually watching/logging what's going on to make it fail and fixing that.

**Comment by ScottG489** [comment]
Hey @chino. I figured out what my issue was!

I finally took a look at the code and noticed that `oathtool` was being imported which surprised me since I was using `oathtool` on my own to get the code. After looking at a few more lines it hit me that I wasn't supposed to be passing in the 6-digit code (in base64) but just the original token itself and the application does the rest for me.

I'm not sure I'd say the documentation should be update at all because looking back at it it does say to use the token. I think I was just confused regarding nomenclature. I had just been playing around with oathtool already so my mind went in another direction and too far ahead.

In any case, your branch helped me get faster testing feedback so I was able to come to this realization :) Thanks again!
